apps/btc_family/btc_txn.c

```diff
@@ -253,24 +236,118 @@ static bool fetch_transaction_meta(btc_query_t *query) {
   memcpy(&g_txn_context->metadata,
          &query->sign_txn.meta,
          sizeof(btc_sign_txn_metadata_t));
-  g_txn_context->inputs = (sign_txn_input_t *)malloc(
-      sizeof(sign_txn_input_t) * g_txn_context->metadata.input_count);
+  g_txn_context->inputs = (btc_txn_input_t *)malloc(
+      sizeof(btc_txn_input_t) * g_txn_context->metadata.input_count);
   g_txn_context->outputs = (btc_sign_txn_output_t *)malloc(
       sizeof(btc_sign_txn_output_t) * g_txn_context->metadata.output_count);
   // TODO: check if malloc failed; report to host and exit
   send_response(BTC_SIGN_TXN_RESPONSE_META_ACCEPTED_TAG);
   return true;
 }
 
+static bool handle_verify_input_utxo(btc_query_t *query) {
+  // Validate input UTXOs for safety from attack. Ref:
+  // https://blog.trezor.io/details-of-firmware-updates-for-trezor-one-version-1-9-1-and-trezor-model-t-version-2-3-1-1eba8f60f2dd
+  for (int idx = 0; idx < g_txn_context->metadata.input_count; idx++) {
+    if (!btc_get_query(query, BTC_QUERY_SIGN_TXN_TAG) &&
+        !check_which_request(query, BTC_SIGN_TXN_REQUEST_INPUT_TAG)) {
+      return false;
+    }
+
+    const btc_sign_txn_input_prev_txn_t *txn = &query->sign_txn.input.prev_txn;
+    if (!btc_verify_utxo(txn->bytes, txn->size, &g_txn_context->inputs[idx])) {
```

Comment:
Please add a comment explaining why input utxo validation is required at this stage (why it can't be decoupled from the usb data flow)